__src__/kalman_filter/EKF/EKF.cpp

```diff
@@ -0,0 +1,142 @@
+    double c21 = 2 * (q(2) * q(3) + q(0) * q(1));
+    double c22 = q(0) * q(0) - q(1) * q(1) - q(2) * q(2) + q(3) * q(3);
+
+    Matrix3d rotMat;
+    rotMat << c00, c01, c02,
+              c10, c11, c12,
+              c20, c21, c22;
+
+    return rotMat;
+}
+
+MatrixXd EKF::getJacobianMatrix(Vector3d reference) {
+    Vector4d qHatPrev = xHatPrev.head<4>();
+    MatrixXd jacobianMatrix(3, 4);
+    jacobianMatrix <<
+        qHatPrev(0) * reference(0) + qHatPrev(3) * reference(1) - qHatPrev(2) * reference(2),
+        qHatPrev(1) * reference(0) + qHatPrev(2) * reference(1) + qHatPrev(3) * reference(2),
+        -qHatPrev(2) * reference(0) + qHatPrev(1) * reference(1) - qHatPrev(0) * reference(2),
+        -qHatPrev(3) * reference(0) + qHatPrev(0) * reference(1) + qHatPrev(1) * reference(2),
+        -qHatPrev(3) * reference(0) + qHatPrev(0) * reference(1) + qHatPrev(1) * reference(2),
+        qHatPrev(2) * reference(0) - qHatPrev(1) * reference(1) + qHatPrev(0) * reference(2),
+        qHatPrev(1) * reference(0) + qHatPrev(2) * reference(1) + qHatPrev(3) * reference(2),
+        -qHatPrev(0) * reference(0) - qHatPrev(3) * reference(1) + qHatPrev(2) * reference(2),
+        qHatPrev(2) * reference(0) - qHatPrev(1) * reference(1) + qHatPrev(0) * reference(2),
+        qHatPrev(3) * reference(0) - qHatPrev(0) * reference(1) - qHatPrev(1) * reference(2),
+        qHatPrev(0) * reference(0) + qHatPrev(3) * reference(1) - qHatPrev(2) * reference(2),
+        qHatPrev(1) * reference(0) + qHatPrev(2) * reference(1) + qHatPrev(3) * reference(2);
+    return 2 * jacobianMatrix;
+}
+
+VectorXd EKF::predictAccelMag() {
+    MatrixXd hPrime_a = getJacobianMatrix(accelReference);
+    Vector3d accelBar = getRotMat(xHatBar.head<4>()).transpose() * accelReference;
+    MatrixXd hPrime_m = getJacobianMatrix(magReference);
+    Vector3d magBar = getRotMat(xHatBar.head<4>()).transpose() * magReference;
+
+    MatrixXd tmp1(3, 7);
+    tmp1 << hPrime_a, MatrixXd::Zero(3, 3);
+    MatrixXd tmp2(3, 7);
+    tmp2 << hPrime_m, MatrixXd::Zero(3, 3);
+    C.resize(6, 7);
+    C << tmp1, tmp2;
+
+    VectorXd result(6);
+    result << accelBar, magBar;
+    return result;
+}
+
+void EKF::predict(Vector3d w, double dt) {
+    Vector4d q = xHat.head<4>();
+    MatrixXd Sq(4, 3);
+    Sq << -q(1), -q(2), -q(3),
+           q(0), -q(3),  q(2),
+           q(3),  q(0), -q(1),
+          -q(2),  q(1),  q(0);
+    MatrixXd tmp1(4, 7);
+    tmp1 << MatrixXd::Identity(4, 4), -dt / 2 * Sq;
+    MatrixXd tmp2(3, 7);
+    tmp2 << MatrixXd::Zero(3, 4), MatrixXd::Identity(3, 3);
+    A.resize(7, 7);
+    A << tmp1, tmp2;
+    B.resize(7, 3);
+    B << dt / 2 * Sq, MatrixXd::Zero(3, 3);
+
+    xHatBar = A * xHat + B * w;
+    xHatBar.head<4>() = normalizeQuat(xHatBar.head<4>());
+    xHatPrev = xHat;
+
+    yHatBar = predictAccelMag();
+    p = A * p * A.transpose() + Q;
+}
+
+void EKF::update(Vector3d a, Vector3d m) {
+    MatrixXd tmp1 = (C * p * C.transpose() + R).inverse();

```

Comment:
I should have known matrix inversion is super expensive! Optimize it next time, no excuses :) This could be a big performance bottleneck. (I love coding).